src/Microsoft.IdentityModel.Tokens/Json/JsonSerializerPrimitives.cs

```diff
@@ -60,194 +61,267 @@ internal static Exception CreateJsonReaderExceptionInvalidType(ref Utf8JsonReade
+        internal static IList<string> ReadStrings(ref Utf8JsonReader reader, IList<string> strings, string propertyName, string className)
         {
+            // returning null keeps the same logic as JsonSerialization.ReadObject
             if (reader.TokenType == JsonTokenType.Null)
-            {
-                if (advanceReader)
-                    reader.Read();
-
                 return null;
-            }
 
-            if (!IsReaderAtTokenType(ref reader, JsonTokenType.StartArray, true))
+            if (!IsReaderAtTokenType(ref reader, JsonTokenType.StartArray, false))
                 throw LogHelper.LogExceptionMessage(
                     CreateJsonReaderExceptionInvalidType(ref reader, "JsonTokenType.StartArray", className, propertyName));
 
-            do
+            while (ReaderRead(ref reader))
             {
-                if (IsReaderAtTokenType(ref reader, JsonTokenType.EndArray, true))
+                if (IsReaderAtTokenType(ref reader, JsonTokenType.EndArray, false))
                     break;
 
-                strings.Add(ReadString(ref reader, propertyName, className, false));
-
-            } while (reader.Read());
+                strings.Add(ReadString(ref reader, propertyName, className));
+            }
 
             return strings;
         }
 
-        // used by JsonWebKey which will be in this release
+        /// <summary>
+        /// This method is only called when we are on a JsonTokenType.Number AND reading into AdditionalData which is an IDictionary[string, object].
+        /// We have to make a choice of the type to return.
+        /// </summary>
+        /// <param name="reader"></param>
+        /// <returns>If possible a .net numerical type, otherwise a JsonElement.</returns>
+        internal static object ReadNumber(ref Utf8JsonReader reader)
+        {
+            // Assume reader is a Utf8JsonReader positioned at a JsonTokenType.Number
+            if (reader.TryGetInt32(out int i))
+                return i;
+            else if (reader.TryGetInt64(out long l))
+                return l;
+            else if (reader.TryGetUInt32(out uint u))
+                return u;
+            else if (reader.TryGetSingle(out float f))
+                return f;
+            else if (reader.TryGetDouble(out double d))
+                return d;
+            else if (reader.TryGetDecimal(out decimal m))
+                return m;
+
+            Debug.Assert(false, "expected to read a number, but none of the Utf8JsonReader.TryGet... methods returned true.");
+
+            return ReadJsonElement(ref reader);
+        }
+
+        internal static void WriteAdditionalData(ref Utf8JsonWriter writer, IDictionary<string, object> additionalData)
+        {
+            if (additionalData.Count > 0)
+            {
+                foreach (string key in additionalData.Keys)
```

Comment:
nit: since you need both the key and value in every iteration it is slightly more efficient to iterate over the KVPs directly and use the key and value off of that

Reply:
@keegan-caruso Yep, make this fix.
Thanks.